sql/rowexec/rel.go

```diff
@@ -732,3 +732,17 @@ func (b *BaseBuilder) buildResolvedTable(ctx *sql.Context, n *plan.ResolvedTable
 func (b *BaseBuilder) buildTableCount(_ *sql.Context, n *plan.TableCountLookup, _ sql.Row) (sql.RowIter, error) {
 	return sql.RowsToRowIter(sql.Row{int64(n.Count())}), nil
 }
+
+func (b *BaseBuilder) buildSlidingRange(ctx *sql.Context, n *plan.SlidingRange, row sql.Row) (sql.RowIter, error) {
```

Comment:
TLDR we want this to construct and return a `RowIter` for the range join that is 1) primed to do work, but hasn't done anything yet, and 2) is ideally a `rowexec` object whose mutations won't impact the original analysis plan objects.
We separate analyzer objects from execution objects. At the end of analysis we have a tree of plan nodes, and that representation is converted into an execution template by the `rowexec` builder. We can make several instances of a single plan execution from the same plan. The template for execution is a volcano row iter, a state machine that will start generating rows when we call `Next()`. The `RowIter` interface, and any implementations (like `filterIter`, `projectIter`, etc) are good examples of how we compartmentalize execution data structures. You could imagine us implementing a `colexec` package in the future, which converts the same analyzer plan to an equivalent but internally different state machine. More practically, this separation makes it easier for me to rewrite the analysis format of the plan, without having to worry about the blast radius reaching the execution format.

Reply:
I see. So I should not be storing execution state in the `plan.SlidingRange` object, because it's just a template for generating plan executions.

This could be tricky, because we need to maintain state (the priority queue of live ranges) in between each time we load and iterate over the secondary table. We'd also like the join iterator to be agnostic about the execution plan of its secondary child.

Do we have other nodes that have this property? Hash joins do some form of caching, right?

Reply:
The trade-off here is maybe (i) if we do not piggy-back off of the default iterator, we have to do more work to add the default null, left join, scope management features to a new `rangeJoinIter` (the same way we do with `mergeJoinIter`, `lateralJoinIter`), or (ii) we will have periodic cache management, isolation, and memory problems with range joins. I personally prefer (i), mostly because (1) our hash join problems have been tricky in the past, (2) I want to rewrite hash join to not use `CachedResults` for performance reasons anyway, and (3) from an architectural perspective I'd rather have execution isolation/speed with new code duplication problems than have analysis and execution interdependent but with speed/cache management problems.

Reply:
Yeah, I can totally see how stateful / primary-row-dependent secondary iters causes caching bugs that are really hard to debug. I'll create a new `rangeJoinIter`.